includes/Pages/UserAuth/Login/LoginCredentialPageBase.php

```diff
@@ -0,0 +1,335 @@
+
+                if ($authResult === AuthenticationManager::AUTH_FAIL) {
+                    throw new ApplicationLogicException("Authentication failed");
+                }
+
+                if ($authResult === AuthenticationManager::AUTH_REQUIRE_NEXT_STAGE) {
+                    $this->processJumpNextStage($user, $partialStage, $database);
+
+                    return;
+                }
+
+                if ($authResult === AuthenticationManager::AUTH_OK) {
+                    $this->processLoginSuccess($user);
+
+                    return;
+                }
+            }
+            catch (ApplicationLogicException $ex) {
+                WebRequest::clearAuthPartialLogin();
+
+                SessionAlert::error($ex->getMessage());
+                $this->redirect('login');
+
+                return;
+            }
+        }
+        else {
+            $this->assign('showSignIn', true);
+
+            $this->setupPartial();
+            $this->assignCSRFToken();
+            $this->providerSpecificSetup();
+        }
+    }
+
+    protected function isProtectedPage()
+    {
+        return false;
+    }
+
+    /**
+     * Enforces HTTPS on the login form
+     *
+     * @return bool
+     */
+    private function enforceHttps()
+    {
+        if ($this->getSiteConfiguration()->getUseStrictTransportSecurity() !== false) {
+            if (WebRequest::isHttps()) {
+                // Client can clearly use HTTPS, so let's enforce it for all connections.
+                $this->headerQueue[] = "Strict-Transport-Security: max-age=15768000";
+            }
+            else {
+                // This is the login form, not the request form. We need protection here.
+                $this->redirectUrl('https://' . WebRequest::serverName() . WebRequest::requestUri());
+
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    protected abstract function providerSpecificSetup();
+
+    protected function setupPartial()
+    {
+        $database = $this->getDatabase();
+
+        // default stuff
+        $this->assign('alternatives', array()); // 'u2f' => array('U2F token'), 'otp' => array('TOTP', 'scratch', 'yubiotp')));
+
+        // is this stage one?
+        list($partialId, $partialStage) = WebRequest::getAuthPartialLogin();
+        if ($partialStage === null || $partialId === null) {
+            WebRequest::clearAuthPartialLogin();
+        }
+
+        // Check to see if we have a partial login in progress
+        $username = null;
+        if ($partialId !== null) {
+            // Yes, enforce this username
+            $this->partialUser = User::getById($partialId, $database);
+            $username = $this->partialUser->getUsername();
+
+            $this->setupAlternates($this->partialUser, $partialStage, $database);
+        }
+        else {
+            // No, see if we've preloaded a username
+            $preloadUsername = WebRequest::getString('tplUsername');
```

Comment:
I haven't tracked where this comes from, but another potential injection point

Reply:
This comes straight from the query string, and is routed to templates/login/shell.tpl, where it's escaped directly for output.